service/raft/raft_group0_client.hh

```diff
@@ -55,22 +64,86 @@ public:
+
+    // `_upgrade_state` is a cached (perhaps outdated) version of the upgrade state stored on disk.
+    group0_upgrade_state _upgrade_state{group0_upgrade_state::recovery}; // loaded from disk in `start()`
+    seastar::rwlock _upgrade_lock;
+    seastar::condition_variable _upgraded;
+
 public:
-    raft_group0_client(service::raft_group_registry& raft_gr) : _raft_gr(raft_gr) {}
+    raft_group0_client(service::raft_group_registry&, db::system_keyspace&);
+
+    // Call after `system_keyspace` is initialized.
+    future<> start();
 
     future<> add_entry(group0_command group0_cmd, group0_guard guard, seastar::abort_source* as = nullptr);
-    bool is_enabled() {
-        return _raft_gr.is_enabled();
-    }
+
+    // Ensures that all previously finished operations on group 0 are visible on this node;
+    // in particular, performs a Raft read barrier on group 0.
+    //
+    // Keep the guard for the entire duration of your operation:
+    // - if the operation requires reading group 0 state (such as schema state), take the guard before doing any read.
+    // - if the operation finishes with appending an entry to the group 0 log, move the guard to `add_entry`.
+    // - if the operation only consists of reads (it does not append any log entry), release the guard
+    //   only after the last read.
+    //
+    // The guard will ensure that given two operations, either:
+    // 1. they won't overlap (where start and end of an operation is defined by the taking and releasing of the guard),
+    // 2. or if they do, one of them will fail (throw `group0_concurrent_modification`) - the application
+    //    of the entry to the group 0 state machine becomes a no-op.
+    //
+    // All successful operations are therefore strictly serialized.
+    //
+    // Call only on shard 0.
+    // FIXME?: this is kind of annoying for the user.
+    // we could forward the call to shard 0, have group0_guard keep a foreign_ptr to the internal data structures on shard 0,
+    // and add_entry would again forward to shard 0.
     future<group0_guard> start_operation(seastar::abort_source* as = nullptr);
 
     group0_command prepare_command(schema_change change, group0_guard& guard, std::string_view description);
 
+    // Returns the current group 0 upgrade state.
+    //
+    // The possible transitions are: `use_old_procedures` -> `synchronize` -> `use_new_procedures`.
+    // Once entering a state, we cannot rollback (even after a restart - the state is persisted).
+    //
+    // An exception to these rules is manual recovery, represented by `recovery` state.
+    // It can be entered by the user manually modifying the system.local table through CQL
+    // and restarting the node. In this state the node will not join group 0 or start the group 0 Raft server,
+    // it will perform all operations as in `use_old_procedures`, and not attempt to perform the upgrade.
```

Comment:
Does this mean we will have to leave the old code forever for "recovery" purposes?

Reply:
To recover the cluster we need to be able to do one of these things:
- remove down nodes from the cluster without group 0,
- or establish group 0 excluding some nodes in the cluster

The proposed recovery procedure uses the first approach: we use the classic `removenode` to remove a down node from the cluster without group 0. Then, after the cluster consists only of live nodes, establish group 0 with the entire cluster.

The second approach would work as follows: establish group 0 without the down node, then remove the remaining traces of the down node using group 0 operations.

Right now there is no disadvantage to the first approach because `removenode` is the only way to remove down nodes from the cluster, we don't have any alternative group 0-specific way.

However, if we implement an alternative way to perform topology changes using group 0, then it will be better to use the second approach, because we won't have to depend on "old code" (by which I mean the classic `removenode`). But that requires some additional interfaces and tools: we will need some way to tell the cluster to create group 0 without a given node, e.g. a tool that performs RPC to each node to start discovery but without a given node.

Reply:
I was not talking about "remove node" part. I was talking about the statement that in `recovery` state old schema change procedure should be use. Ideally we should be able to remove the old procedure after two versions with raft enabled (I know it is not possible to do it that fast because of the enterprise, but you get the idea).

Reply:
> after two versions with raft enabled

my understanding is:
- ~two versions where raft is enabled but optional
- ~two versions where raft is mandatory
- only then we will be able to remove the old code

so ~4 versions in total

> I was talking about the statement that in recovery state old schema change procedure should be use.

We can disable schema changes during recovery altogether, if we want to. So we won't need the "schema change code".

But we'll need to have some way to ensure that the existing schema is the same on all nodes during recovery before we create group 0. Whether it's going to use the same code as now or some different code, I don't know.
But I think we'll keep reusing most of the existing code - all that code with merging schema mutations etc. will stay - the only thing that will become "redundant" maybe is stuff like fetching schema by RPC, although that can be used to ensure schema is synchronized, so maybe it will stay useful.

Reply:
> But we'll need to have some way to ensure that the existing schema is the same on all nodes during recovery before we create group 0. 

So about that. Strictly speaking yes, we should. But in practice I am not sure this is required. Since we still pull the schema even if stated are not synchronized they eventually will be. I think using this fact could have simplified the upgrade code as well. I do not see why can't we allow two ways of doing schema change *in parallel*. Sure the result at this point will not be linearizable, but we do not needed it to be at this point.   

> Whether it's going to use the same code as now or some different code, I don't know. But I think we'll keep reusing most of the existing code - all that code with merging schema mutations etc. will stay - the only thing that will become "redundant" maybe is stuff like fetching schema by RPC, although that can be used to ensure schema is synchronized, so maybe it will stay useful.

The code that pushes schema can go away for sure.

Reply:
> > But we'll need to have some way to ensure that the existing schema is the same on all nodes during recovery before we create group 0.
> 
> So about that. Strictly speaking yes, we should. But in practice I am not sure this is required. Since we still pull the schema even if stated are not synchronized they eventually will be. I think using this fact could have simplified the upgrade code as well. I do not see why can't we allow two ways of doing schema change _in parallel_. Sure the result at this point will not be linearizable, but we do not needed it to be at this point.
> 
> > Whether it's going to use the same code as now or some different code, I don't know. But I think we'll keep reusing most of the existing code - all that code with merging schema mutations etc. will stay - the only thing that will become "redundant" maybe is stuff like fetching schema by RPC, although that can be used to ensure schema is synchronized, so maybe it will stay useful.
> 
> The code that pushes schema can go away for sure.

I think not waiting for the schema to sync during the upgrade will violate linearizability of schema changes on Raft. Of course we could say that such schema changes are eventually linearizable, but then we won't be able to say *when* this eventuality will happen.
Imagine we elect a Raft leader with an outdated schema. Since we don't force schema sync, it may accept group0 entries based on an outdated schema state. And later will fetch a "newer" schema hash through the migration manager and update the local schema hash. Given that the schema hash is also part of state_id history table, we'll end up in a mess. 

I'd rather avoid it.

Reply:
> > > But we'll need to have some way to ensure that the existing schema is the same on all nodes during recovery before we create group 0.
> > 
> > > Whether it's going to use the same code as now or some different code, I don't know. But I think we'll keep reusing most of the existing code - all that code with merging schema mutations etc. will stay - the only thing that will become "redundant" maybe is stuff like fetching schema by RPC, although that can be used to ensure schema is synchronized, so maybe it will stay useful.
> > 
> > 
> > The code that pushes schema can go away for sure.
> 
> I think not waiting for the schema to sync during the upgrade will violate linearizability of schema changes on Raft. Of course we could say that such schema changes are eventually linearizable, but then we won't be able to say _when_ this eventuality will happen. Imagine we elect a Raft leader with an outdated schema. Since we don't force schema sync, it may accept group0 entries based on an outdated schema state. And later will fetch a "newer" schema hash through the migration manager and update the local schema hash. Given that the schema hash is also part of state_id history table, we'll end up in a mess.

We do not accept or reject based on schema state, only history table state which is not part of the state pushed by old schema code and, if accessed properly, will be the same on all group0 participants.

Reply:
Right, but will it provide linearizable semantics? The group0 state machine will accept changes based on an outdated state.

Reply:
I mean the whole read-modify-write semantics of group0 client will go bonkers.

Reply:
So the patch does three things basically: freezes old-style schema updates, syncs schema, and moves to group zero. Later  I left a comment that we could do it in a fewer number of steps, you here suggest that the order of steps could be different, good at least we agree what needs to be done :)

Reply:
> Since we still pull the schema even if stated are not synchronized they eventually will be.

Unless schema is in a state that prevents it from being synchronized.
Then we end up in a state where we upgraded to group 0, but fixed none of the problems we were supposed to fix.

The current algorithm will refuse to upgrade if schema is impossible to synchronize, forcing the user to fix the problem (if there's any) and guarantees that if upgrade finishes, we end up in consistent schema with linearizable changes.